pkg/chains/formats/intotoite6/extract/extract_test.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2022 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package extract_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/tektoncd/chains/pkg/chains/formats/intotoite6/extract"
+	"github.com/tektoncd/chains/pkg/chains/objects"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	logtesting "knative.dev/pkg/logging/testing"
+)
+
+const (
+	artifactURL1    = "gcr.io/test/kaniko-chains1"
+	artifactDigest1 = "a2e500bebfe16cf12fc56316ba72c645e1d29054541dc1ab6c286197434170a9"
+	artifactURL2    = "us-central1-maven.pkg.dev/test/java"
+	artifactDigest2 = "b2e500bebfe16cf12fc56316ba72c645e1d29054541dc1ab6c286197434170a9"
+)
+
+var (
+	tr = objects.NewTaskRunObject(
+		&v1beta1.TaskRun{
+			Status: v1beta1.TaskRunStatus{
+				TaskRunStatusFields: v1beta1.TaskRunStatusFields{
+					TaskRunResults: []v1beta1.TaskRunResult{
+						{Name: "y_IMAGE_DIGEST", Value: *v1beta1.NewStructuredValues("sha256:" + artifactDigest1)},
+						{Name: "y_IMAGE_URL", Value: *v1beta1.NewStructuredValues(artifactURL1)},
+						{Name: "x_ARTIFACT_DIGEST", Value: *v1beta1.NewStructuredValues("sha256:" + artifactDigest2)},
+						{Name: "x_ARTIFACT_URI", Value: *v1beta1.NewStructuredValues(artifactURL2)},
+					},
+				},
+			},
+		},
+	)
+
+	pr = objects.NewPipelineRunObject(
+		&v1beta1.PipelineRun{
+			Status: v1beta1.PipelineRunStatus{
+				PipelineRunStatusFields: v1beta1.PipelineRunStatusFields{
+					PipelineResults: []v1beta1.PipelineRunResult{
+						{Name: "y_IMAGE_DIGEST", Value: *v1beta1.NewStructuredValues("sha256:" + artifactDigest1)},
+						{Name: "y_IMAGE_URL", Value: *v1beta1.NewStructuredValues(artifactURL1)},
+						{Name: "x_ARTIFACT_DIGEST", Value: *v1beta1.NewStructuredValues("sha256:" + artifactDigest2)},
+						{Name: "x_ARTIFACT_URI", Value: *v1beta1.NewStructuredValues(artifactURL2)},
+					},
+				},
+			},
+		},
+	)
+)
+
+func TestSubjectDigests(t *testing.T) {
+	var tests = []struct {
+		name         string
+		runObj       objects.TektonObject
+		wantSubjects []intoto.Subject
+	}{
```

Comment:
I'd add some test cases where there are malformed / missing digests as well. e.g.

- Result with name but no result value
- sha1 (or any other non-sha256)
- Hash without `sha256:` prefix
- whatever else you can think of!

Reply:
Good points. I've added all of them including some other edge cases i.e. typo in type hinting field name, invalid sha256 value etc. PTAL! Thanks!